File: core/src/androidTest/mock/com/mobilesolutionworks/android/controller/test/util/HostAndHostController.java
//package com.mobilesolutionworks.android.controller.test.util;
//
//
//import com.mobilesolutionworks.android.app.controller.HostWorksController;
//import com.mobilesolutionworks.android.controller.test.GetController;
//
//import static org.junit.Assert.assertEquals;
//import static org.junit.Assert.assertNotEquals;
//import static org.junit.Assert.assertNotNull;
//
///**
// * Created by yunarta on 15/3/17.
// */
//
//public class HostAndHostController<C extends HostWorksController> {
//
//    private boolean mChangeHost;
//
//    private C mController;
//
//    private int mHostHash;
//
//    private int mControllerHash;
//
//    public HostAndHostController() {
//        this(true);
//    }
//
//    public HostAndHostController(boolean changeHost) {
//        mChangeHost = changeHost;
//    }
//
//    public void set(GetController<C> host) {
//        assertNotNull(host);
//
//        mController = host.getController();
//        assertNotNull(mController);
//
//        mHostHash = System.identityHashCode(host);
//        mControllerHash = System.identityHashCode(mController);
//    }
//
//    public void validate(GetController<C> host) {
//        assertNotNull(host);
//        assertNotNull(host.getController());
//
//        if (mChangeHost) {
//            assertNotEquals("Host changed after rotation", mHostHash, System.identityHashCode(host));
//        } else {
//            assertEquals("Host didn't changed after rotation", mHostHash, System.identityHashCode(host));
//        }
//
//        assertEquals("Controller is not retain during rotation", mControllerHash, System.identityHashCode(host.getController()));
//
//        assertEquals("Host in controller is not the same with host that creates", host, mController.getHost());
//    }
//
//    public C getController() {
//        return mController;
//    }
//}

File: test-app/src/androidTest/kotlin/com/mobilesolutionworks/android/controller/test/util/HostAndHostController.kt
package com.mobilesolutionworks.android.controller.test.util


import com.mobilesolutionworks.android.app.controller.HostWorksController
import com.mobilesolutionworks.android.controller.test.GetController
import org.junit.Assert.*

/**
 * Created by yunarta on 15/3/17.
 */

class HostAndHostController<C : HostWorksController<*>> constructor(private val mChangeHost: Boolean = true) {

    var controller: C? = null
        private set

    private var mHostHash: Int = 0

    private var mControllerHash: Int = 0

    fun set(host: GetController<C>) {
        assertNotNull(host)

        controller = host.controller
        assertNotNull(controller)

        mHostHash = System.identityHashCode(host)
        mControllerHash = System.identityHashCode(controller)
    }

    fun validate(host: GetController<C>) {
        assertNotNull(host)
        assertNotNull(host.controller)

        if (mChangeHost) {
            assertNotEquals("Host changed after rotation", mHostHash.toLong(), System.identityHashCode(host).toLong())
        } else {
            assertEquals("Host didn't changed after rotation", mHostHash.toLong(), System.identityHashCode(host).toLong())
        }

        assertEquals("Controller is not retain during rotation", mControllerHash.toLong(), System.identityHashCode(host.controller).toLong())

        assertEquals("Host in controller is not the same with host that creates", host, controller!!.host)
    }
}
